### PR TITLE
Domain type

### LINKED ISSUE #7316: Support domains for Postgresql
### Describe the use case

Postgresql supports domains using the [`CREATE DOMAIN`](https://www.postgresql.org/docs/current/sql-createdomain.html) and [`DROP DOMAIN`](https://www.postgresql.org/docs/current/sql-dropdomain.html) commands. A domain is a data type with additional constraints -- for example, an email address is a specific kind of string, so you can use a domain to define a type that only accepts strings that are correctly formatted as email addresses.

I would like to be able to define and manage domains using SQLAlchemy, in the same kind of way that I am already able to define and manage enums with SQLAlchemy.

### Databases / Backends / Drivers targeted

Postgresql only. As far as I can tell, no other database that SQLAlchemy supports, supports domains.

### Example Use

```python
import sqlalchemy as sa

Email = Domain(
    name="email",
    data_type=sa.Text,
    constraint=sa.CheckConstraint(r"VALUE ~ '[^@]+@[^@]+\.[^@]+'"),
)
t1 = sa.Table(
    "table",
    metadata,
    sa.Column("id", sa.Integer, primary_key=True),
    sa.Column("email", Email),
)
```

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by singingwolfboy**
### Description
Fixes #7316. I mostly based this code off the code for `Enum`. I wasn't able to get the code to work with the `CheckConstraint` class, so I ended up just allowing the developer to pass the constraint as a plain string. I'm also very hesitant on the automated tests; I added a few, but I'm sure there's more that should be added. As for documentation, I mostly just copied the docstrings for `Enum`, although I'm sure there's more documentation needed.

Please consider this PR as a starting point. If you can give me some guidance on how to proceed, I'm happy to try to push it forward to a point that it can be merged. Alternatively, if one of the maintainers is interested in finishing this up without me, I'm fine with that as well!

### Checklist

This pull request is:

- [ ] A documentation / typographical error fix
	- Good to go, no issue or tests are needed
- [ ] A short code fix
	- please include the issue number, and create an issue if none exists, which
	  must include a complete example of the issue.  one line code fixes without an
	  issue and demonstration will not be accepted.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.   one line code fixes without tests will not be accepted.
- [x] A new feature implementation
	- please include the issue number, and create an issue if none exists, which must
	  include a complete example of how the feature would look.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.

### CODE REVIEW

**Comment by CaselIT** [comment]
Hi,

From a fist look this seems like a great start, thanks!
I've asked mike to take a look. In the meantime I'll import the code in gerrit, since postgres is missing from the workflow here on github

**Review by sqla-tester — COMMENTED** [review]
OK, this is **sqla-tester** setting up my work on behalf of **CaselIT** to try to get revision bb6fbe3b39e989718e36ac6dd622cff088c7bb80 of this pull request into gerrit so we can run tests and reviews and stuff

**Comment by sqla-tester** [comment]
New Gerrit review created for change bb6fbe3b39e989718e36ac6dd622cff088c7bb80: https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3305

File: lib/sqlalchemy/dialects/postgresql/base.py [hunk 1]
```diff
@@ -2041,6 +2041,167 @@ def _on_metadata_drop(self, target, bind, checkfirst=False, **kw):
...
+         :class:`_engine.Connection`, or similar object to emit
+         SQL.
+        :param checkfirst: if ``True``, a query against
+         the PG catalog will be first performed to see
+         if the type actually exists before dropping.
+
+        """
+        bind._run_ddl_visitor(self.DomainDropper, self, checkfirst=checkfirst)
+
+    def _check_for_name_in_memos(self, checkfirst, kw):
```
Comment (zzzeek) [inline lib/sqlalchemy/dialects/postgresql/base.py:2108]: if we go with this, all the code here that's copied from ENUM needs to be factored out into a common base.  I dont want all this duplication.

**Comment by CaselIT** [comment]
Thanks for the update. It looks better without the duplication.

Did you already find a way of reflecting the domains from the db? Just the sql string is fine, it's ok is it's not yet in sqlalchemy

**Comment by singingwolfboy** [comment]
> Did you already find a way of reflecting the domains from the db? Just the sql string is fine, it's ok is it's not yet in sqlalchemy

I haven't looked into reflection yet; I don't know much about that. So I suppose the answer to your question is no. How would I go about doing that?

**Comment by CaselIT** [comment]
> 
> 
> > Did you already find a way of reflecting the domains from the db? Just the sql string is fine, it's ok is it's not yet in sqlalchemy
> 
> I haven't looked into reflection yet; I don't know much about that. So I suppose the answer to your question is no. How would I go about doing that?

usually by searching for it to find if someone had the same issue and there is an sql query for that :)

It seems that there is not much to add to the existing code, just a reflection of the check constraint, that can be added by adding a left join on `pg_constraint` on `contypid` and getting the definition with `pg_catalog.pg_get_constraintdef(pg_constraint."oid")` https://www.postgresql.org/docs/11/catalog-pg-constraint.html

All in all seems pretty doable.

@zzzeek considering also alembic and the changes done in the multi reflection patch that is in progress, I think this may be a 2.0 thing

**Comment by singingwolfboy** [comment]
> It seems that there is not much to add to the existing code, just a reflection of the check constraint, that can be added by adding a left join on `pg_constraint` on `contypid` and getting the definition with `pg_catalog.pg_get_constraintdef(pg_constraint."oid")` https://www.postgresql.org/docs/11/catalog-pg-constraint.html
> 
> All in all seems pretty doable.

I'm glad that it seems doable, but with my very small amount of knowledge around SQLAlchemy's internals and around Postgresql's `pg_catalog` tables, I don't think I'm up for this task. I'm not sure what data we need to pull, how to write the SQL query, where in the codebase it belongs, what else might be affected by this change, and so on. @CaselIT, are you able to make this change, since you seem to understand what needs to be done here?

**Comment by singingwolfboy** [comment]
@CaselIT, can you please help me out with the reflection thing? As I said, I don't know what I'm doing here. I want to keep pushing this pull request forward, rather than let it get stale and die, but I'm not sure what else I can do here, since.

**Comment by CaselIT** [comment]
I'm finishing a couple of large changes, but I'll try looking into it next.

Thanks for the work so far!

**Review by sqla-tester — COMMENTED** [review]
OK, this is **sqla-tester** setting up my work on behalf of **CaselIT** to try to get revision a3d75a704906ae68d4a884467d33da259e433347 of this pull request into gerrit so we can run tests and reviews and stuff

**Comment by sqla-tester** [comment]
Patchset a3d75a704906ae68d4a884467d33da259e433347 added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3305

**Comment by CaselIT** [comment]
> I'm finishing a couple of large changes, but I'll try looking into it next.
> 
> Thanks for the work so far!

Sorry, but I've yet to have time to look at this, I'm not ignoring it.

**Review by sqla-tester — COMMENTED** [review]
OK, this is **sqla-tester** setting up my work on behalf of **CaselIT** to try to get revision 719e93b042147b6bf01439077e3d6535e7e6d8c7 of this pull request into gerrit so we can run tests and reviews and stuff

**Comment by sqla-tester** [comment]
Patchset 719e93b042147b6bf01439077e3d6535e7e6d8c7 added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3305

**Comment by singingwolfboy** [comment]
> if this can be rebased to latest main branch we can run it through CI again and suggest fixes for the failing tests

Done!

**Review by sqla-tester — COMMENTED** [review]
OK, this is **sqla-tester** setting up my work on behalf of **zzzeek** to try to get revision bc9a82f010e6ca2f70a6e8a7620b748e483c26c3 of this pull request into gerrit so we can run tests and reviews and stuff

**Comment by sqla-tester** [comment]
Patchset bc9a82f010e6ca2f70a6e8a7620b748e483c26c3 added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3305

**Comment by sqla-tester** [comment]
**mike bayer** (zzzeek) wrote:

this looks like it's pretty good, the only failure is a test suite not dropping an ENUM for some reason.   If you aren't able to debug that particular failure I can go in and figure out what it is.     the change here would also need a changelog file added in doc/build/changelog/unreleased_20/, we can write those pretty quickly but feel free if you get to it first.

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3305

**Comment by sqla-tester** [comment]
**mike bayer** (zzzeek) wrote:

as you can see the goal here is to unify ENUM/DOMAIN as much as possible, with the exception that ENUM is also doing "native" for the Enum datatype.  Otherwise, ENUM/DOMAIN need to be as interchangeable as possible both for API clarity as well as maintainability.

this is close to being ready but DOMAIN needs a lot more docs.

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3305

**Comment by singingwolfboy** [comment]
I definitely feel like I need some help with debugging test failures, writing doc strings, and so on. I wanted to make this pull request as a starting point, but I'm not very familiar with this codebase, and I'll accept any help I can get to get this change over the finish line 😄

**Comment by sqla-tester** [comment]
**mike bayer** (zzzeek) wrote:

> Patch Set 6:
> 
> singingwolfboy@github wrote:
> 
> I definitely feel like I need some help with debugging test failures, writing doc strings, and so on. I wanted to make this pull request as a starting point, but I'm not very familiar with this codebase, and I'll accept any help I can get to get this change over the finish line 😄

yeah no worries, I put it out there for anyone who wants to help w/ that but ultimately I always go over all the docs in any case before merging, when I get back to this I will add some more in.

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3305

**Review by sqla-tester — COMMENTED** [review]
OK, this is **sqla-tester** setting up my work on behalf of **CaselIT** to try to get revision bc9a82f010e6ca2f70a6e8a7620b748e483c26c3 of this pull request into gerrit so we can run tests and reviews and stuff

**Comment by sqla-tester** [comment]
Failed to create a gerrit review, git squash against branch 'main' failed

**Review by sqla-tester — COMMENTED** [review]
**Federico Caselli** (CaselIT) wrote:

I've added test for the domain reflection. I've also added reflection of their constraint, since it seemed pointless not to support it.

Next I'm gonna rebase this on to of the multi reflection.

Mike the -1 is on the get_all method that as is I don't think it's great.

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3305

* lib/sqlalchemy/dialects/postgresql/named_types.py (line 31): what's the rationale on this get_all method? I'm -1 on it at the moment, since no other element has something similar in sqlalchemy.

also we get back a list of dict not a list of domani / enum, so I don't think we should have it, or at least not as currently implemented

**Review by sqla-tester — COMMENTED** [review]
**Federico Caselli** (CaselIT) wrote:

This should be final, but it should go on top of multi reflection

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3305

* lib/sqlalchemy/dialects/postgresql/named_types.py (line 31): Done

**Review by sqla-tester — COMMENTED** [review]
**mike bayer** (zzzeek) wrote:

looks totally great.   a few nits but also I think we should have a quick example doc b.c. right now i wouldn't know how to use this.

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3305

* doc/build/changelog/unreleased_20/7316.rst (line 5): PostgreSQL
* lib/sqlalchemy/dialects/postgresql/base.py (line 1438): since this is for backwards compat I would export these:

    from .named_types import CreateDomainType as CreateDomainType

etc., the noqas can be removed
* lib/sqlalchemy/dialects/postgresql/base.py (line 2808): use an f string
* lib/sqlalchemy/dialects/postgresql/named_types.py (line 349): maybe an example here?   start with "e.g.::"
* lib/sqlalchemy/dialects/postgresql/named_types.py (line 373): these have to be done for this to be useful....

**Review by sqla-tester — COMMENTED** [review]
**Federico Caselli** (CaselIT) wrote:

will work on in once I rebase this on top of multi reflection

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3305

* lib/sqlalchemy/dialects/postgresql/named_types.py (line 349): makes sense
* lib/sqlalchemy/dialects/postgresql/named_types.py (line 373): I had not noticed them. will do

**Review by sqla-tester — COMMENTED** [review]
**Federico Caselli** (CaselIT) wrote:

Ok I modified the api to be more similar to postgresql definition

I've also used the compiler functionality to render the user provided strings

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3305

* doc/build/changelog/unreleased_20/7316.rst (line 5): Ack
* lib/sqlalchemy/dialects/postgresql/base.py (line 1438): Ack
* lib/sqlalchemy/dialects/postgresql/base.py (line 2808): Done
* lib/sqlalchemy/dialects/postgresql/named_types.py (line 349): Done
* lib/sqlalchemy/dialects/postgresql/named_types.py (line 373): Done

**Comment by sqla-tester** [comment]
**Federico Caselli** (CaselIT) wrote:

Ok I've changed a bit the api of the domain to be more similar to the create  domain of pg.

Can you take another look?

NOTE: this is now on top of multi reflection

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3305

**Comment by sqla-tester** [comment]
**Federico Caselli** (CaselIT) wrote:

Can you take a second look?

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3305

**Comment by sqla-tester** [comment]
Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3305 has been **merged**. Congratulations! :)

**Pull request closed without merging**